test/libsolidity/interface/FileReader.cpp

```diff
@@ -0,0 +1,443 @@
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+/// Unit tests for libsolidity/interface/FileReader.h
+
+#include <libsolidity/interface/FileReader.h>
+
+#include <test/Common.h>
+#include <test/FilesystemUtils.h>
+#include <test/TemporaryDirectory.h>
+#include <test/libsolidity/util/SoltestErrors.h>
+
+#include <boost/filesystem.hpp>
+#include <boost/test/unit_test.hpp>
+
+using namespace std;
+using namespace solidity::test;
+
+
+namespace solidity::frontend::test
+{
+
+BOOST_AUTO_TEST_SUITE(FileReaderTest)
+
+BOOST_AUTO_TEST_CASE(normalizeCLIPathForVFS_absolute_path)
+{
+	BOOST_TEST(FileReader::normalizeCLIPathForVFS("/") == "/");
+	BOOST_TEST(FileReader::normalizeCLIPathForVFS("/.") == "/");
+	BOOST_TEST(FileReader::normalizeCLIPathForVFS("/./") == "/");
+	BOOST_TEST(FileReader::normalizeCLIPathForVFS("/./.") == "/");
+
+	BOOST_TEST(FileReader::normalizeCLIPathForVFS("/a") == "/a");
+	BOOST_TEST(FileReader::normalizeCLIPathForVFS("/a/") == "/a/");
+	BOOST_TEST(FileReader::normalizeCLIPathForVFS("/a/.") == "/a/");
+	BOOST_TEST(FileReader::normalizeCLIPathForVFS("/./a") == "/a");
+	BOOST_TEST(FileReader::normalizeCLIPathForVFS("/./a/") == "/a/");
+	BOOST_TEST(FileReader::normalizeCLIPathForVFS("/./a/.") == "/a/");
+	BOOST_TEST(FileReader::normalizeCLIPathForVFS("/a/b") == "/a/b");
+	BOOST_TEST(FileReader::normalizeCLIPathForVFS("/a/b/") == "/a/b/");
+
+	BOOST_TEST(FileReader::normalizeCLIPathForVFS("/a/./b/") == "/a/b/");
+	BOOST_TEST(FileReader::normalizeCLIPathForVFS("/a/../a/b/") == "/a/b/");
+	BOOST_TEST(FileReader::normalizeCLIPathForVFS("/a/b/c/..") == "/a/b");
+	BOOST_TEST(FileReader::normalizeCLIPathForVFS("/a/b/c/../") == "/a/b/");
+
+	BOOST_TEST(FileReader::normalizeCLIPathForVFS("/a/b/c/../../..") == "/");
+	BOOST_TEST(FileReader::normalizeCLIPathForVFS("/a/b/c/../../../") == "/");
+}
+
+BOOST_AUTO_TEST_CASE(normalizeCLIPathForVFS_relative_path)
+{
+	TemporaryDirectory tempDir("file-reader-test-");
+	boost::filesystem::create_directories(tempDir.path() / "x/y/z");
+	TemporaryWorkingDirectory tempWorkDir(tempDir.path() / "x/y/z");
+
+	// NOTE: If path to work dir contains symlinks (of then the case on macOS), boost might resolve
+	// them, making the path different from tempDirPath.
+	boost::filesystem::path expectedWorkDir = boost::filesystem::current_path().parent_path().parent_path().parent_path();
```

Comment:
Why parent 3 times?

Reply:
Because the working directory is set to `tempDir.path() / "x/y/z"` and the paths I'm testing below sometimes go up a few levels so they do not always preserve the `x/y/z` part. So I basically want `tempDir.path()` but normalized the way a working directory would be (e.g. on MacOS the temp dir I get usually has a symlink in it that I need to resolve; on Windows it has a drive letter).

But I just realized that now it's no longer working dir so the name is a bit misleading so maybe that's what you meant. Renamed to `expectedPrefix`.